File: go/chapter.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

// Chapter represents a learning chapter loaded from JSON
type Chapter struct {
	ID            string     `json:"id"`
	Title         string     `json:"title"`
	Description   string     `json:"description"`
	Theory        string     `json:"theory"`
	Exercises     []Exercise `json:"exercises"`
	ExerciseCount int        `json:"exercise_count"`
}

// Exercise represents a single exercise in a chapter
type Exercise struct {
	ID             string   `json:"id"`
	Title          string   `json:"title"`
	Description    string   `json:"description"`
	StarterCode    string   `json:"starter_code"`
	ExpectedOutput string   `json:"expected_output"`
	Hints          []string `json:"hints"`
	Solution       string   `json:"solution"`
}

// ChapterLoader loads chapters from JSON files
type ChapterLoader struct {
	chaptersDir string
}

// NewChapterLoader creates a new chapter loader
func NewChapterLoader(chaptersDir string) *ChapterLoader {
	return &ChapterLoader{chaptersDir: chaptersDir}
}

// LoadChapters loads all chapters from the chapters directory
func (cl *ChapterLoader) LoadChapters() ([]Chapter, error) {
	// Read directory
	entries, err := os.ReadDir(cl.chaptersDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read chapters directory: %w", err)
	}

	var paths []string
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		if filepath.Ext(entry.Name()) != ".json" {
			continue
		}
		paths = append(paths, entry.Name())
	}
	sort.Strings(paths)

	var chapters []Chapter
	for _, name := range paths {
		chapter, err := cl.loadChapterFile(filepath.Join(cl.chaptersDir, name))
		if err != nil {
			return nil, fmt.Errorf("failed to load %s: %w", name, err)
		}
		chapters = append(chapters, chapter)
	}

	if len(chapters) == 0 {
		return nil, fmt.Errorf("no chapters found in %s", cl.chaptersDir)
	}

	return chapters, nil
}

// loadChapterFile loads a single chapter from a JSON file
func (cl *ChapterLoader) loadChapterFile(path string) (Chapter, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return Chapter{}, fmt.Errorf("failed to read file: %w", err)
	}

	var chapter Chapter
	if err := json.Unmarshal(data, &chapter); err != nil {
		return Chapter{}, fmt.Errorf("failed to parse JSON: %w", err)
	}

	// Validate required fields
	if chapter.ID == "" {
		return Chapter{}, fmt.Errorf("missing required field: id")
	}
	if chapter.Title == "" {
		return Chapter{}, fmt.Errorf("missing required field: title")
	}
	if chapter.Description == "" {
		return Chapter{}, fmt.Errorf("missing required field: description")
	}

	// Validate exercises
	if len(chapter.Exercises) == 0 {
		return Chapter{}, fmt.Errorf("chapter has no exercises")
	}

	// Validate each exercise
	for i, ex := range chapter.Exercises {
		if ex.ID == "" {
			return Chapter{}, fmt.Errorf("exercise %d: missing id", i)
		}
		if ex.Title == "" {
			return Chapter{}, fmt.Errorf("exercise %d (%s): missing title", i, ex.ID)
		}
		if ex.StarterCode == "" {
			return Chapter{}, fmt.Errorf("exercise %d (%s): missing starter_code", i, ex.ID)
		}
		if ex.ExpectedOutput == "" {
			return Chapter{}, fmt.Errorf("exercise %d (%s): missing expected_output", i, ex.ID)
		}
		// Solution is optional but recommended
	}

	// Set exercise count
	chapter.ExerciseCount = len(chapter.Exercises)

	return chapter, nil
}

// LoadChapter loads a single chapter by ID
func (cl *ChapterLoader) LoadChapter(id string) (*Chapter, error) {
	chapters, err := cl.LoadChapters()
	if err != nil {
		return nil, err
	}

	for i := range chapters {
		if chapters[i].ID == id {
			return &chapters[i], nil
		}
	}

	return nil, fmt.Errorf("chapter not found: %s", id)
}

// GetDefaultChaptersDir returns the default chapters directory
func GetDefaultChaptersDir() string {
	return "chapters"
}

File: rust/src/chapter.rs
use serde::Deserialize;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Deserialize)]
pub struct Chapter {
    pub id: String,
    pub title: String,
    pub description: String,
    pub theory: String,
    pub exercises: Vec<Exercise>,
    #[serde(default)]
    pub exercise_count: usize,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Exercise {
    pub id: String,
    pub title: String,
    pub description: String,
    #[serde(rename = "starter_code")]
    pub starter_code: String,
    #[serde(rename = "expected_output")]
    pub expected_output: String,
    #[serde(default)]
    pub hints: Vec<String>,
    #[serde(default)]
    pub solution: String,
}

pub struct ChapterLoader {
    chapters_dir: PathBuf,
}

impl ChapterLoader {
    pub fn new(chapters_dir: impl Into<PathBuf>) -> Self {
        Self {
            chapters_dir: chapters_dir.into(),
        }
    }

    pub fn load_chapters(&self) -> Result<Vec<Chapter>, String> {
        let entries = fs::read_dir(&self.chapters_dir)
            .map_err(|e| format!("failed to read chapters directory: {e}"))?;

        let mut paths: Vec<PathBuf> = entries
            .filter_map(|e| e.ok())
            .map(|e| e.path())
            .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("json"))
            .collect();

        paths.sort();

        if paths.is_empty() {
            return Err(format!(
                "no chapters found in {}",
                self.chapters_dir.display()
            ));
        }

        let mut chapters = Vec::new();
        for path in paths {
            chapters.push(self.load_chapter_file(&path)?);
        }

        Ok(chapters)
    }

    fn load_chapter_file(&self, path: &Path) -> Result<Chapter, String> {
        let data = fs::read_to_string(path).map_err(|e| format!("failed to read file: {e}"))?;

        let mut chapter: Chapter =
            serde_json::from_str(&data).map_err(|e| format!("failed to parse JSON: {e}"))?;

        if chapter.id.is_empty() {
            return Err("missing required field: id".into());
        }
        if chapter.title.is_empty() {
            return Err("missing required field: title".into());
        }
        if chapter.description.is_empty() {
            return Err("missing required field: description".into());
        }
        if chapter.exercises.is_empty() {
            return Err("chapter has no exercises".into());
        }

        for (i, ex) in chapter.exercises.iter().enumerate() {
            if ex.id.is_empty() {
                return Err(format!("exercise {i}: missing id"));
            }
            if ex.title.is_empty() {
                return Err(format!("exercise {i} ({}): missing title", ex.id));
            }
            if ex.starter_code.is_empty() {
                return Err(format!("exercise {i} ({}): missing starter_code", ex.id));
            }
            if ex.expected_output.is_empty() {
                return Err(format!(
                    "exercise {i} ({}): missing expected_output",
                    ex.id
                ));
            }
        }

        chapter.exercise_count = chapter.exercises.len();
        Ok(chapter)
    }

    /// Load a single chapter by `id` (must match the `id` field inside the JSON file).
    #[allow(dead_code)]
    pub fn load_chapter(&self, id: &str) -> Result<Chapter, String> {
        self.load_chapters()?
            .into_iter()
            .find(|c| c.id == id)
            .ok_or_else(|| format!("chapter not found: {id}"))
    }
}

pub fn default_chapters_dir() -> PathBuf {
    PathBuf::from("chapters")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn load_chapters_smoke() {
        let loader = ChapterLoader::new("chapters");
        let chapters = loader.load_chapters().expect("load chapters");
        assert!(
            chapters.len() >= 19,
            "expected at least 19 chapters, got {}",
            chapters.len()
        );
    }

    #[test]
    fn load_chapter_by_id() {
        let loader = ChapterLoader::new("chapters");
        let ch = loader.load_chapter("variables").expect("variables");
        assert_eq!(ch.id, "variables");
        assert!(!ch.exercises.is_empty());
    }

    #[test]
    fn load_missing_chapter_errors() {
        let loader = ChapterLoader::new("chapters");
        assert!(loader.load_chapter("nonexistent").is_err());
    }
}
